data/transactions/logic/eval.go

```diff
@@ -841,10 +846,27 @@ func eval(program []byte, cx *EvalContext) (pass bool, err error) {
 
 	if cx.Tracer != nil {
 		cx.Tracer.BeforeProgram(cx)
+
 		defer func() {
+			if _, ok := err.(PanicError); ok {
+				// A panic error occurred during the eval loop. Report it now.
+				cx.Tracer.AfterOpcode(cx, err)
+			}
+
 			// Ensure we update the tracer before exiting
 			cx.Tracer.AfterProgram(cx, err)
 		}()
+
+		// Due to the LIFO behavior of defer statements, if we want the tracer to catch panics, we
+		// need to have a `recover` defer statement defined after the tracer-invoking defer
+		// statement. However, since it's possible for the tracer itself to panic, we still want to
+		// keep the outmost `recover` defer statement defined at the top of this function.
+		defer func() {
+			if x := recover(); x != nil {
+				err = makePanicError(x, cx.EvalParams, "Eval")
+				pass = false
+			}
+		}()
```

Comment:
Definitely looking for feedback on this. As the comment mentions, go defer statements are executed LIFO, so without this additional recovery, the above defer which invokes tracer methods will not see any PanicErrors, as those are only created by the defer at the top of this function.

By adding another defer layer that can recover from panics, we make them visible to the tracer.

An alternative here would be to just move the single recovery defer from the beginning of the function to after we define this tracer deferred logic. I opted against it since I didn't want to decrease coverage of potential panicking code, especially since it's in theory possible for tracer methods themselves to panic. There's a new test case, `TestEvalWithTracerPanic`, which confirm a tracer method panicking gets caught during eval.

Reply:
I think I follow what's going on, but let me ask in order to be sure.  You are trying to catch the panic "early", and build the stack trace so that you have it available to give out to the tracer.  Might you simply call the Tracer with a simpler error here, and then panic again with `x`?

That is, only one `defer`, something like:
```
		defer func() {
           x := recover();
           if x != nil {
				// A panic error occurred during the eval loop. Report it now.
				cx.Tracer.AfterOpcode(cx, fmt.Errorf("some nonsense"))
			}

			// Ensure we update the tracer before exiting
			cx.Tracer.AfterProgram(cx, err)

            if x != nil {
                panic(x)
            }
		}()
```

I recognize this gives `AfterOpcode` slightly less information to go on. But I'm trying to make it so that a panic is always handled through the single stack trace creating handler. Done this way, when there's a Tracer, we build the stack here, and we _don't_ go through the existing panic handler, since we're no longer panicking.

My motivation is that I think I can see a future where we really want to preserve the difference of "normal" error returns from evaluation vs a panic, if/when we consider supporting  a `itxn_try_submit` which I think should continue to fail for panics, rather than cleanly reporting 0 for a failed txn.

Reply:
Good idea. I implemented it in 912d5de3d, let me know if you see any issues